src/crawlee/basic_crawler/_basic_crawler.py

```diff
@@ -865,6 +865,9 @@ async def __run_task_function(self) -> None:
             else:
                 self._logger.exception('Request failed and reached maximum retries', exc_info=session_error)
 
+                if self._error_handler:
+                    await self._error_handler(crawling_context, session_error)
+
```

Comment:
This is different from what the JS version does - the error handler should be called before retries, not after all attempts have failed.